Pcap++/src/PcapFilter.cpp

```diff
@@ -38,95 +27,176 @@ namespace pcpp
+	}
+
+	bool BpfFilterWrapper::matches(const RawPacket& rawPacket) const
+	{
+		return matches(rawPacket.getRawData(), rawPacket.getRawDataLen(), rawPacket.getPacketTimeStamp(),
+		               rawPacket.getLinkLayerType());
+	}
+
+	bool BpfFilterWrapper::matches(const uint8_t* packetData, uint32_t packetDataLength, timespec timestamp,
+	                               uint16_t linkType) const
 	{
 		if (m_FilterStr.empty())
 			return true;
 
-		if (!setFilter(std::string(m_FilterStr), static_cast<LinkLayerType>(linkType)))
+		// This should never happen, but just in case
+		if (m_CachedProgram == nullptr)
 		{
-			return false;
+			throw std::runtime_error("No compiled BPF program available");
+		}
+
+		// Handle link type mismatch
+		if (linkType != static_cast<uint16_t>(m_CachedProgramLinkType))
+		{
+			auto newProgram = compileFilter(m_FilterStr, static_cast<LinkLayerType>(linkType));
+			if (newProgram == nullptr)
+			{
+				PCPP_LOG_ERROR("Couldn't compile BPF filter: '" << m_FilterStr << "' for link type: " << linkType);
+				return false;
+			}
+			m_CachedProgram = std::move(newProgram);
+			m_CachedProgramLinkType = static_cast<LinkLayerType>(linkType);
 		}
 
-		struct pcap_pkthdr pktHdr;
+		// Test the packet against the filter
+		pcap_pkthdr pktHdr;
 		pktHdr.caplen = packetDataLength;
 		pktHdr.len = packetDataLength;
-		pktHdr.ts = internal::toTimeval(packetTimestamp);
+		pktHdr.ts = internal::toTimeval(timestamp);
+		return (pcap_offline_filter(m_CachedProgram.get(), &pktHdr, packetData) != 0);
+	}
+
+	BpfFilterWrapper::BpfProgramUPtr BpfFilterWrapper::compileFilter(std::string const& filter, LinkLayerType linkType)
+	{
+		if (filter.empty())
+			return nullptr;
+
+		auto pcap = std::unique_ptr<pcap_t, internal::PcapCloseDeleter>(pcap_open_dead(linkType, DEFAULT_SNAPLEN));
+		if (pcap == nullptr)
+		{
+			return nullptr;
+		}
+
+		auto newProg = std::make_unique<bpf_program>();
+		int ret = pcap_compile(pcap.get(), newProg.get(), filter.c_str(), 1, 0);
+		if (ret < 0)
+		{
+			return nullptr;
+		}
+
+		// Reassigns ownership to a new unique_ptr with a custom deleter as it now requires specialized cleanup.
+		return BpfProgramUPtr(newProg.release());
+	}
+
+	bool GeneralFilter::matchPacketWithFilter(RawPacket* rawPacket)
+	{
+		if (rawPacket == nullptr)
+		{
+			PCPP_LOG_ERROR("Raw packet pointer is null");
+			return false;
+		}
+
+		return matches(*rawPacket);
+	}
+
+	bool GeneralFilter::matches(RawPacket const& rawPacket) const
+	{
+		if (!m_CachedFilter)
+		{
+			if (!cacheFilterInternal())
+			{
+				return false;
+			}
+		}
 
-		return (pcap_offline_filter(m_Program.get(), &pktHdr, packetData) != 0);
+		return m_BpfWrapper.matches(rawPacket);
+	}
+
+	bool GeneralFilter::cacheFilterInternal() const
+	{
+		std::string filterStr;
+		parseToString(filterStr);
+		if (!m_BpfWrapper.setFilter(filterStr))
+			return false;
+
+		m_CachedFilter = true;
+		return true;
 	}
```

Comment:
Why do we need this method? `BpfFilterWrapper::setFilter()` already prevents resetting the filter if it hasn't changed

Reply:
I don't think recomputing the filter string is a big deal since it is not done in real time